docs/foreign-keys.md

```diff
@@ -0,0 +1,199 @@
+# ADR 3: Extending the `relationship` field to allow Record-to-Record references 
+
+Jean Cochrane (Open Source Fellow, Summer 2018)
+
+## Context
+
+In [#32](https://github.com/azavea/grout-2018-fellowship/issues/32),
+I introduced a proposal to extend the `relationship` field to allow references
+to other Records in the datastore. This ADR investigates whether that feature
+is possible, and proposes next steps for it.
+
+### Background
+
+The basic idea of a Record-to-Record reference is to permit one Record to
+point to another. For example, imagine two RecordTypes that store data about
+Events as well as the advertisements for those events, Posters:
+
+- Poster
+    - `image`: Image
+    - `date`: DateTime
+    - `location`: Coordinates
+
+- Event
+    - `title`: String
+    - `date`: DateTime
+    - `location`: Coordinates
+
+A Record-to-Record reference would allow us to link Posters to the Events
+that they represent:
+
+- Poster
+    - `image`: Image
+    - `date`: DateTime
+    - `location`: Coordinates
+    - **`event`**: **UUID**
+
+Then, Posters could be grouped, searched, and filtered according to the Event
+that they reference.
+
+#### Implementation description
+
+Currently, references are implemented under the `localReference` data type in
+the schema editor (titled `Relationships`) and only permit references to
+fields contained within the referencing Record itself.
+
+In the schema editor, Record-to-Record foreign keys could be represented in a new field, such as
+`externalReference`, with a schema definition along these lines:
+
+```json
+"externalReference": {
+    "allOf": [
+        { "$ref": "#/definitions/abstractBaseField" },
+        { "$ref": "#/definitions/abstractRequirableField" }
+    ],
+    "title": "Record-to-Record Relationship",
+    "properties": {
+        "fieldTitle": {
+            "title": "Relationship Title",
+            "type": "string",
+            "minLength": 1
+        },
+        "referenceTarget": {
+            "title": "RecordType of the related Record",
+            "type": "string",
+            "format": "select",
+            "enumSource": [["Will be dynamically overwritten"]]
```

Comment:
I like the `$ref` pattern you sketched out a lot -- really nice that it can express the idea in pure JSONSchema!

It seems like you're suggesting implementing a kind of limited `ON DELETE` standard, which would make a lot of sense. At the very least, allowing the equivalents of `CASCADE` and `SET NULL` seems reasonable to me.

I still don't quite understand the validation issue, however. My understanding is that a Record only gets validated when it enters the database (via `POST`) or is updated (via `PUT`). To me, it seems desirable to complain loudly in those two cases if a reference doesn't resolve properly. When else might we expect validation to get triggered?

Reply:
My initial concern with validation was that we currently have as an invariant that `jsonschema.validate(record.schema.schema, record.data)` (made-up function) is always true. This would make that no longer an invariant because some parts of the schema would be dynamic based on database state, but I didn't have a more concrete worry than that. Now that you've narrowed things down to creation and update (which I think is correct), I think that my primary concern is about the update case -- it doesn't seem very nice to allow a record to silently get into an invalid state, such that an unsuspecting user could try to make an edit that was unrelated to the external reference and find themselves receiving validation errors. I think that having `ON DELETE` equivalents would still solve the issue, however.

Reply:
This may be a case where the JSON schema standard hasn't quite caught up with the usage we want, but it's [under consideration for draft 4](https://spacetelescope.github.io/understanding-json-schema/structuring.html). I also stumbled across some good discussion here: json-schema-org/json-schema-spec#514 regarding delegation versus inclusion for reference targets.

Reply:
I think that's probably okay then; it looks like they're up to Draft 7 now: http://json-schema.org/specification-links.html#draft-4 . It's very confusing that they just call everything drafts; I'm unclear whether any of these have been finalized.

Reply:
That *is* confusing. The `json-editor` library README claims compatibility with both versions 3 and 4, so it's unclear where they intend to stand on reference behavior then.

As an aside, I just noticed [a Vue JSON schema validator exists](https://github.com/mokkabonna/vue-vuelidate-jsonschema). It's linked to in the schema docs, [here](http://json-schema.org/implementations.html). Maybe that would be helpful to look at for some of the front-end work.

Reply:
The PR that [closed the delegation-vs-inclusion issue](json-schema-org/json-schema-spec#628) was just merged eighteen days ago, so this aspect does seem to be under current development.